### ISSUE TITLE
Implementing "spanning edge betweenness" for edge reliability calculation in MST

### ISSUE THREAD

**Issue by dnieuw** [igraph/igraph#1376]
Dear Igraph team,

I'm working on a nursing home and hospital transmission study of SARS-CoV-2 using your igraph package to create and calculate with graphs. My current results are based on the creation of interaction networkds by using genome sequence and patient data and creating an MST for cluster analysis. I have been looking for a way to calculate edge reliability in MSTs and found this recent thesis describing "spanning edge betweenness" as a measure of edge reliability (chapter 3.1):

https://www.researchgate.net/profile/Andreia_Teixeira8/publication/334707019_Complex_networks_analysis_from_an_edge_perspective/links/5d3f44b292851cd04691ddd0/Complex-networks-analysis-from-an-edge-perspective.pdf
(I am not involved in this research in any way)

This seems a very interesting and broadly applicable measure, but currently, it is implemented only in java (as far as I know) and it is available in a tool called phyloviz. I wonder if it would be possible to implement this measure into igraph? It would be an extremely useful tool to estimate the confidence level of edges in MST analyses, and I think it solves a major issue in the interpretation of MST which can prove useful in a variety of fields.

**Comment by ntamas** [comment]
Are there any performance problems with the Java implementation? If there aren't any, a quick workaround would be to write the edge list from igraph into a simple edge list file (one edge per line, source-target vertex IDs), call the Java implementation, and parse the results. It is probably easier than reimplementing the algorithm in C, and time is the essence when it comes to SARS-CoV-2-related analyses. (We could probably create an efficient implementation in C, but it could be a matter of weeks if we want to do it "properly").

So, if there are no big performance problems with the Java version in general (or you can afford throwing more computational power at the problem), let me know and we can help with building the glue code that calls the Java implementation from igraph.

**Comment by dnieuw** [comment]
I understand, stupid to not think about that first, I'll have a look at the execution speed of the algorithm. In the thesis, she mentions that small networks take seconds but large networks can take an hour, so have to see how that translates to my data. She indicates that it has O(n^2.5) complexity in the worst case.

FYI, I'm currently using the R interface to your C core to run igraph, if you could help with the "glue" that would be great.

For future reference, this is the java implementation she has written. I do not know how easy this would be to port to C as I'm not literate in either of them.
https://bitbucket.org/phyloviz/popsim-analysis/src/master/src/net/phyloviz/msn/TreeStatsGeneric.java

The method consumes a tsv edge list file and a max node id and outputs the same edge list and spanning edge betweenness value as a space-separated list to stdout, so that should be easy enough to parse/glue.

Thanks for the lightning-fast response!

**Comment by dnieuw** [comment]
The algorithm runs for about an hour on a 444 vertex fully connected network on 4 cores, which is quite slow, but doable.

The issue is though that the "generic network" java implementation does not support weighted MSTs, which is crucial for the network I have, which is generated from an all vs all sequence distance matrix. Therefore the results assigned a very low probability to all edges since in a fully connected unweighted network the MST is basically a random connection of vertices. This will most likely also have impacted the performance.

A (sub-optimal) temporary solution would be to set a weight cutoff for the edges and use that trimmed but unweighted network for the MST and spanning edge betweenness calculation. But that could give strange results and again an underestimation of edge probability.

I will try to contact the author and see if she can help me out, but again I would be very interested in this metric for future projects and I think that this kind of edge reliability would be a great addition to the igraph toolkit.

**Comment by szhorvat** [comment]
Can you give a concise, but complete and mathematically precise definition of this measure you requested?

**Comment by dnieuw** [comment]
I am not a mathematician myself so I'm sorry if this is not what you were asking for, but this is how she describes the algorithm in her thesis:

```
Given a connected, undirected and weighted graph G = (V, E), with weight function w : E →
IR, we can compute the number of MSTs and the spanning edge betweenness for each edge as follows:
1. sort E with respect to w in increasing order;
2. let H = (V, ∅) and τG = 1;
3. iterate over E in increasing order and, while edges have the same weights, add them to H;
4. for each connected component C in H, compute τC using Theorem 1, update τG = τG × τC , and,
for each edge e ∈ C, compute τC (e) using Theorem 2 and δC (e) using Equation 3.1;
5. contract all edges in H such that each connected component becomes a single vertex;
6. if H has more than one vertex, repeat from step 3, otherwise return τG.
```
The Theorems and Equation 3.1 do not lend themselves well for posting here because of the mathematical formulas they contain, they can be found described in detail in her thesis in chapter 3.1:
https://www.researchgate.net/publication/324504497_Models_and_large-scale_analysis_of_dynamics_in_complex_networks_from_biological_to_social_systems

The metric has been described in her publication here:
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4370493/

It has been cited/used/discussed in 12 citing publications since.

**Comment by ASTeixeira** [comment]
Dear all, I'm happy to help. 

I was the one developing it and I even created a Python notebook with the same calculations: https://github.com/ASTeixeira/SpanningEdgeBetwenness/

Regarding publications there are already three publications with the measure as you can see in the GitHub repository. So it was already peer reviewed and has been used. The original publication "Spanning Edge Betweenness" has 22 citations. 

It would be my pleasure to help to integrate Spanning Edge Betweenness in iGraph. 
And the java version indeed supports weighted networks. The input file should be an edge list with in the format: source dest weight

Let me know if/how can I help.

**Comment by szhorvat** [comment]
@ASTeixeira Welcome, and thanks for the offer to help!

I have to apologize in advance for not having read the paper yet, but I am a bit overwhelmed at the moment. I was hoping that you could clarify a few things, to make this go faster.

I'm looking at http://snap.stanford.edu/mlg2013/submissions/mlg2013_submission_14.pdf

> This metric is defined as the fraction of  minimum  spanning  trees  where  a  given  edge  is  present

Also looking at the unweighted part of your Python code, basically you count spanning trees in the graph using Kirchhoff's theorem, then count those which contain an edge `e` using a similar theorem, then finally take the ratio of the two.

Is this correct?

One concern I have about using Kirchhoff's theorem is that it will give us enormous numbers. I wonder if there are numerical accuracy issues when computing the determinant using `double`s, and if overflow is a concern.

I just generated a random connected graph with 100 vertices and 200 edges (i.e. not very large). It had 31767852322155199689055006023697153505497809503 spanning trees. That's `3.17 * 10^46`. A `double` only goes up to about `10^308`.

Do you have a better method for computing the ratio than computing both the numerator and the denominator explicitly?

----

Regarding the weighted case:

If the edge weights are real numbers, I would expect most weighted graphs to have precisely one minimum spanning tree, thus the spanning edge betweenness would be trivially 1 for any edge in the tree and zero for others. Am I correct in thinking that this measure is primarily useful in cases where the weights are small integers?

----

> Let me know if/how can I help.

A Java implementation cannot be integrated into igraph, as igraph is written in C and C++. If you would like to contribute a C implementation to igraph, it would be _very_ welcome, and I'd be happy to guide you through any technical difficulties. Unfortunately, I do not have the time to implement it on my own. (I cannot speak for other regular contributors, but we're pretty short on manpower ...)

**Comment by ntamas** [comment]
@ASTeixeira Could you please point me towards the implementation that handles weighted graphs? I was looking at the [code on Bitbucket](https://bitbucket.org/phyloviz/popsim-analysis/src/99c006a95b974664a9cd866e72dfb039bba6bff3/src/net/phyloviz/msn/TreeStatsGeneric.java#lines-51), but it seems to ignore the third column in the input file. Is there an alternative implementation I should look at?

**Comment by ASTeixeira** [comment]
> Also looking at the unweighted part of your Python code, basically you count spanning trees in the graph using Kirchhoff's theorem, then count those which contain an edge e using a similar theorem, then finally take the ratio of the two.
> 
> Is this correct?

Yes it is. We take a row and columns in the Laplacian for the total number of MSTs and then we take rows and columns of a given edge to obtain the number of MSTs the it is part of. Then we do the ratio. 

> Do you have a better method for computing the ratio than computing both the numerator and the denominator explicitly?

Hum. I'll have to think about this. For the multiplications in weighted networks I have, and I actually have to update the Python repository. I can do the calculation in a log way, so I have sum of log values and in the end we just have to put 10^to the value. It is how is implemented in the Java version. Because I needed only to test small examples I forgot to put that, but I will very soon. I'll think better about this. For sparse graphs this shouldn't be a problem, but for dense...

> Regarding the weighted case:
> 
> If the edge weights are real numbers, I would expect most weighted graphs to have precisely one minimum spanning tree, thus the spanning edge betweenness would be trivially 1 for any edge in the tree and zero for others. Am I correct in thinking that this measure is primarily useful in cases where the weights are small integers?

I think it has to do with the heterogeneity of the weights. If the heterogeneity if very high, then most probably there will be only one MST, yes. For more flat weight distributions (as it can be the case in phylogeny) you get more for sure. 

> A Java implementation cannot be integrated into igraph, as igraph is written in C and C++. If you would like to contribute a C implementation to igraph, it would be very welcome, and I'd be happy to guide you through any technical difficulties. Unfortunately, I do not have the time to implement it on my own. (I cannot speak for other regular contributors, but we're pretty short on manpower ...)

These are weird times and I understand completely. Let me try to figure out a strategy to try to get this done.

**Comment by ASTeixeira** [comment]
> @ASTeixeira Could you please point me towards the implementation that handles weighted graphs? I was looking at the [code on Bitbucket](https://bitbucket.org/phyloviz/popsim-analysis/src/99c006a95b974664a9cd866e72dfb039bba6bff3/src/net/phyloviz/msn/TreeStatsGeneric.java#lines-51), but it seems to ignore the third column in the input file. Is there an alternative implementation I should look at?

Oh, I see.  I checked. This must have been a weird update of the repository. 
This:
`				int src = new Integer(s[0]);
				int target = new Integer(s[1]);
				EdgeMST edge = new EdgeMST(src, target, 1);	`

should be:
`int src = Integer.parseInt(s[0]);
				int target = Integer.parseInt(s[1]);
				int weight = Integer.parseInt(s[2]);
				EdgeMST edge = new EdgeMST(src, target, weight);	`

I'm not sure what happened with the repository. I'm sorry. I'll correct that. 

Meanwhile there is python notebook version here: https://github.com/ASTeixeira/SpanningEdgeBetwenness/

It should give the same results, but I'm still improving the implementation.

**Comment by szhorvat** [comment]
> These are weird times and I understand completely. Let me try to figure out a strategy to try to get this done.

So basically what you need to replicate your *current* implementation in C is a library that is able to compute the determinants of sparse matrices efficiently. Or better: the logarithm of the determinant. Is this correct?

**Comment by szhorvat** [comment]
@ntamas Determinant (or log-determinat) calculations for sparse matrices should be available in the part of SuiteSparse that's included in igraph, right? I've never worked with SuiteSparse. Do you have any experience with this?

**Comment by ASTeixeira** [comment]
> > These are weird times and I understand completely. Let me try to figure out a strategy to try to get this done.
> 
> So basically what you need to replicate your _current_ implementation in C is a library that is able to compute the determinants of sparse matrices efficiently. Or better: the logarithm of the determinant. Is this correct?

Yes. The rest are a bunch of data structures (Disjoint Sets) and so on that have to be used to implement each step of the weighted graphs. For unweighted is pretty straitforward.

**Comment by ntamas** [comment]
> This must have been a weird update of the repository.

Thanks, with your modification it seems to work now!

I think we are dealing with two separate issues here. One is the inclusion of this measure into igraph; the other one is helping @dnieuw to be able to use this measure from his code that is using igraph (from R, I guess) - assuming that he is willing to quantize the edge weights in his networks to reduce heterogeneity, otherwise it is very likely that he will end up with a single MST.

I think that this topic should primarily be used for discussing the eventual implementation of this measure in igraph. As for making a quick "bridge" from R to the Java-based code, that's a different story and is only tangentially related to igraph. @dnieuw - can you drop me an email at ntamas dot gmail dot com so we can discuss this further?

**Comment by szhorvat** [comment]
@dnieuw Please check the comments above on there being usually only one minimum spanning tree in a general weighted graph. You only gave a very short description of your problem, but based on that I am quite sceptical about whether this measure is suitable for your use case ...

Something closely related that I tried on a few occasions was randomly perturbing the weights slightly and looking at how often each edge is present in the MST depending on the amount of perturbation. This can be done on a graph having real-valued "heterogeneous" weights.

**Comment by ASTeixeira** [comment]
@ntamas  The repository should be updated now: https://bitbucket.org/phyloviz/popsim-analysis/src/master/src/net/phyloviz/msn/TreeStatsGeneric.java 
Thank you for pointing the problem!

Regarding @dnieuw work I've applied this in phylogeny analysis before. If the weights are distances between sequences, it should be ok. In the end it would be a complete graph with integer distances representing how different the sequences are. Actually phylogeny analysis was the first motivation to develop the centrality measure, to find confidence in the branches of the phylogenetic tree generated by any algorithm. I'm available to help, for sure.

**Comment by dnieuw** [comment]
Great to see your responses here.

Regarding your questions about the nature of the distances, the distances are integer values representing mutation counts and indel events so no need to quantize. I can, however, see some application of floating-point weights in phylogenetic analyses, such as with evolutionary model based sequence distances.

Regarding the applicability, since the weights are often the same between different strains of the virus (many strains are only a few mutations apart) my prediction is that there are indeed many MSTs possible. On the contrary, if there is only 1 MST possible and I can prove that by calculating the spanning edge betweenness that would be great because that would give us evidence that the infection route in the shown MST is the correct route. So either way, SEB would be very useful.

I'll contact you @ntamas for the "glue" part. If you need any help with the R implementation, I am a skilled R programmer so would be happy to help.
